kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java

```diff
@@ -188,4 +212,56 @@ private void waitRandomTime() throws InterruptedException {
     // wait
     Thread.sleep(retryWaitMs + random.nextInt(WAIT_MAX_JITTER));
   }
+
+  /**
+   * Filter and send Records to DLQ according to allowedBigQueryReason List
+   * @param rows The rows to write.
+   * @param failedRowsMap map from failed row id to the BigQueryError
+   * @param table The BigQuery table to write the rows to
+   * @return Filtered Map from failed row id to the BigQueryError which cannot be sent to DLQ
+   */
+  private Map<Long, List<BigQueryError>> filterAndSendRecordsToDLQ(
+      SortedMap<SinkRecord, InsertAllRequest.RowToInsert> rows,
+      Map<Long, List<BigQueryError>> failedRowsMap,
+      PartitionedTableId table
+  ) {
+    long index = 0;
+    Set<SinkRecord> recordsToDLQ = new TreeSet<>(rows.comparator());
+    Map<Long, List<BigQueryError>> recordsToDLQFailureMap = new TreeMap<>();
+    Map<Long, List<BigQueryError>> updatedFailedRowMap = new TreeMap<>();
+    for (Map.Entry<SinkRecord, InsertAllRequest.RowToInsert> row : rows.entrySet()) {
+      if (failedRowsMap.containsKey(index)) {
+        if (errantRecordHandler.isErrorReasonAllowed(failedRowsMap.get(index).get(0).getReason())) {
```

Comment:
Is checking only 1st element intentional ? `failedRowsMap.get(index).get(0).getReason()`

Reply:
`failedRowsMap.get(index)` will return List<BigQueryError> for that particular failed row. As per my understanding - A row will failed with single error reason only and thus we are checking for 1st element.

Reply:
Yeah, I would assume only a single error reason but was unsure why would it return a list. This should be fine if we already tested it.

Reply:
Let me check if I can find something in documentation related to this - why it is returning a list of error. That would clear the confusion.

Reply:
https://cloud.google.com/java/docs/reference/google-cloud-bigquery/latest/com.google.cloud.bigquery.InsertAllResponse#com_google_cloud_bigquery_InsertAllResponse_getInsertErrors__ 
Do not see anything on when we can expect multiple entries in list.

Reply:
@aniketshrimal , did you find anything related for this ? If not, lets make sure we test this piece out.

Reply:
@b-goyal I could not find anything related to this as well. Let me know if this approach sounds good to you - Instead of checking for first element - lets iterate over error list and if any of the failure reason matches the `allowedBigQueryErrorReason` list - then we will send the record to dlq otherwise we will throw the exception.

Reply:
As per offline discussion - When all the errors in list are in allowedBigQueryErrorReason list - then only we will sent the record to dlq.